Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +86,295 @@ extension CertificatePolicy {
+            // Success
+            if result == 1 { return result }
+
+            // Custom error handling
+            let errorCode = CCryptoBoringSSL_X509_STORE_CTX_get_error(ctx)
+            // Certs could have unknown critical extensions and cause them to be rejected.
+            // Instead of disabling all critical extension checks with X509_V_FLAG_IGNORE_CRITICAL
+            // we will just ignore this specific error.
+            if errorCode == X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION {
+                return 1
+            }
+            return result
+        }
+        CCryptoBoringSSL_X509_STORE_CTX_set_verify_cb(x509StoreCtx, verifyCallback)
+
+        guard CCryptoBoringSSL_X509_verify_cert(x509StoreCtx) == 1 else {
+            let error = CCryptoBoringSSL_X509_verify_cert_error_string(numericCast(CCryptoBoringSSL_X509_STORE_CTX_get_error(x509StoreCtx)))
+            diagnosticsEngine.emit(warning: "The certificate is invalid: \(String(describing: error))")
+            return wrappedCallback(.failure(CertificatePolicyError.invalidCertChain))
+        }
+
+        if certChain.count >= 1, let httpClient = httpClient {
+            // Whether cert chain can be trusted depends on OCSP result
+            ocspClient.checkStatus(certificate: certChain[0], issuer: certChain[1], httpClient: httpClient,
+                                   diagnosticsEngine: diagnosticsEngine, callbackQueue: callbackQueue, callback: callback)
+        } else {
+            wrappedCallback(.success(()))
+        }
     }
+    #endif
 }
 
+#if os(Linux) || os(Windows)
+private let ocspClient = BoringSSLOCSPClient()
+
+private struct BoringSSLOCSPClient {
+    private let resultCache = ThreadSafeKeyValueStore<CacheKey, CacheValue>()
+
+    private let cacheTTL: DispatchTimeInterval
+
+    init(cacheTTL: DispatchTimeInterval = .seconds(300)) {
+        self.cacheTTL = cacheTTL
+    }
+
+    func checkStatus(certificate: Certificate,
+                     issuer: Certificate,
+                     httpClient: HTTPClient,
+                     diagnosticsEngine: DiagnosticsEngine,
+                     callbackQueue: DispatchQueue,
+                     callback: @escaping (Result<Void, Error>) -> Void) {
+        let wrappedCallback: (Result<Void, Error>) -> Void = { result in callbackQueue.async { callback(result) } }
+
+        let ocspURLs = CCryptoBoringSSL_X509_get1_ocsp(certificate.underlying)
+        defer { CCryptoBoringSSL_sk_OPENSSL_STRING_free(ocspURLs) }
+
+        let ocspURLCount = CCryptoBoringSSL_sk_OPENSSL_STRING_num(ocspURLs)
+        // Nothing to do if no OCSP URLs. Use `supportsOCSP` to require OCSP support if needed.
+        guard ocspURLCount > 0 else { return wrappedCallback(.success(())) }
+
+        // Construct the OCSP request
+        let digest = CCryptoBoringSSL_EVP_sha1()
+        guard let certid = OCSP_cert_to_id(digest, certificate.underlying, issuer.underlying),
```

Comment:
`certid` is heap located and needs to be freed.

Reply:
Since `certid` is ultimately a part of `request`, wouldn't freeing `request` take care of `certid` too?

Reply:
Yes, you're right, if the immediate next usage is to pass to a function labelled `add0` then we are handing ownership to that new type. However, there is still a failure mode here: if the call to `OCSP_request_add0_id` fails then we will leak this pointer.

Reply:
`OCSP_request_add0_id` frees `OCSP_ONEREQ` in case of error:
https://github.com/apple/swift-package-manager/pull/3272/files#diff-9f556dcdb9de0622179dc1aa4fd297a26a83fe0ea78a99c60f4bd5de78e3e90dR50

My assumption is that `OCSP_ONEREQ` has ownership of `certid` at that point, so `certid` should be freed as well?

Reply:
Not if it returned NULL from OCSP_ONEREQ_new. This is actually a nasty part of this API: there appear to be two paths out, one of which calls a free and the other does not. This can only be correct if OCSP_ONEREQ does _not_ take ownership of the pointer. Unfortunately, because this is openssl, it's basically impossible to know if it _is_ correct.

In this instance we can assume that failure path is impossible (we can't survive OOM conditions in Swift) so I think for now we can take the hit. Worst case we leak: not the end of the world in this kind of tool.

Reply:
> Not if it returned NULL from OCSP_ONEREQ_new

Ah true.

> I think for now we can take the hit. Worst case we leak: not the end of the world in this kind of tool.

Yes, I would rather take the hit than crashing while trying to free `certid`. Also since the CLI just runs and exits (i.e., not long running) the impact is minimal.